expanding piggy bank tutorial (#500)
Concordium/concordium.github.io: source/mainnet/smart-contracts/tutorials/piggy-bank/deploying.rst

.. code-block:: console

   $cargo concordium build

.. Note::
   This command creates a module ``piggy_bank_part2.wasm.v1`` in the folder ``./target/concordium/wasm32-unknown-unknown/release/``. Please use the wasm module file suffixed with `v1` which is the newest smart contract version supported by the Concordium chain.

The ``piggy_bank_part2.wasm.v1`` file can be deployed to the Concordium chain. Let’s transfer the ``piggy_bank_part2.wasm.v1`` file to the place where you are running your concordium-client tool. Move to the folder as well.

Deploying a module
==================
You are now set up to deploy the piggy bank module to the testnet chain.

.. Note::
   Please remember to adjust the following commands when you are running your ``concordium-client`` tool locally and want to connect to your node on the server. You can look up the different options on how to connect the ``concordium-client`` tool to your testnet node :ref:`here<interacting_with_your_testnet_node>`.

.. code-block:: console

>    $./concordium-client module deploy piggy_bank_part2.wasm.v1 --sender <Account name> --grpc-port 10001

Reviewer: I would use the name parameter here, `--name piggy_bank_part2_module`.

Reviewer: We are not completely consistent elsewhere in the documentation, but we mostly use `<snake-case>` for placeholders in the smart contract parts. So please use that :)

.. Note::
   This tutorial uses ``Concordium`` for the <Account name>.

.. Note::
   You will be asked to input a password. It is the password that you use for your mobile app wallet log-in.

If everything works correctly the output has a green line with your module reference. Remember this module reference since it will be used in the next step. You can find additional information about deploying a module in this :ref:`guide<deploy-module>`.

Initializing a module
=====================

You can initialize your smart contract now to create a smart contract instance.

.. code-block:: console

>    $./concordium-client contract init <module reference> --sender <Account name> --contract PiggyBank --energy 1000 --grpc-port 10001

Reviewer: I would use names again. 
I usually name my modules and contract identically, but that might confuse new users, so consider using some naming scheme to distinguish them.

>    $./concordium-client contract init <module reference> --sender <Account name> --contract PiggyBank --energy 1000 --grpc-port 10001

Author: I call them `piggy_bank_part2_instance` and `piggy_bank_part2_module` in the tutorial now.

.. Note::
   Use enough energy to initialize the piggy bank smart contract.

If everything works correctly the output has a green line with your smart contract instance index. Remember this index since it will be used in the next step.

.. code-block:: console

   Contract successfully initialized with address: {"index":<Smart contract instance index>,"subindex":0}

You can find additional information about initializing a smart contract instance in this :ref:`guide<initialize-contract>`.

Updating the piggy bank smart contract
======================================

Let's deposit 1 CCD into the piggy bank to test if we can update the smart contract instance.

.. code-block:: console

>    $./concordium-client contract update <Smart contract instance index> --entrypoint insert --energy 1000 --sender <Account name> --amount 1 --grpc-port 10001

Reviewer: Please make sure that the specified max energy is sufficient. Also when initializing the contract.

If everything works correctly the output has a green line as follows:

.. code-block:: console

>    Successfully updated contract instance {"index":<Smart contract instance index>,"subindex":0} using the function 'insert'.

Reviewer: ```suggestion
   Successfully updated contract instance {"index":<Smart contract instance index>,"subindex":0}(piggy_bank_part2) using the function 'insert'.
```
I believe this is how contracts with names are formatted. But please check to make sure.

You can find additional information about updating a smart contract instance in this :ref:`guide<interact-instance>`.

Viewing the piggy bank state
============================

> You can display the raw state of your piggy bank smart contract with the following command.

Author: I mean with "raw" state that it is not pretty-printed into a JSON output like it is when using the `schema.bin` file. I suggest adding another smart contract example in the future or expanding this example one day, so can show how to display state values with a `schemas` file.

.. code-block:: console

>    $./concordium-client contract invoke <Smart contract instance index> --entrypoint view --energy 1000 --grpc-port 10001

Reviewer: Use the name.

Reviewer: I think you should show the output of the view function.

> You can find additional information about invoking in this :ref:`guide<invoke-instance>`.

Reviewer: For completeness, I think you should also smash the contract to collect the CCD. It will show the user that not all functions expect amounts (the smash will even reject if you send it CCD, so there should be a note about that).

.. _Rust: https://www.rust-lang.org/

.. _piggy-bank-deploying:

=======================================
Deploying the piggy bank smart contract
=======================================

This is the fourth :ref:`part of a tutorial<piggy-bank>` on smart contract
development.
So far you have written and tested a piggy bank smart contract in the Rust_ programming language.

> Furthermore, you set up your testnet node on a server, created an account that is funded with some CCD, and  imported your account key into the ``concordium-client`` configuration.

Reviewer: ```suggestion
Furthermore, you set up your testnet node on a server, created an account that is funded with some CCD, and imported your account key into the ``concordium-client`` configuration.
```

This part will focus on how you can deploy your developed piggy bank smart contract to the Concordium testnet and interact with it.

.. warning::

   The reader is assumed to have basic knowledge of what a blockchain and smart
   contract is, and some experience with Rust_.

Preparation
===========

Before you start, make sure you completed the previous parts of the tutorial and have a running testnet node that has caught up to the head of the blockchain and an account with some CCD setup.

Building a wasm module
======================

You have already built a wasm module from your piggy bank smart contract in a :ref:`previous part<cargo-concordium-build>` of this tutorial.

Let's repeat this step since you might have done some changes to the smart contract code in :ref:`part 2<piggy-bank-testing>` of this tutorial. Move to the folder where you saved your piggy bank smart contract and use ``cargo concordium build`` to create a wasm module.

.. code-block:: console

   $cargo concordium build

.. Note::
   This command creates a module ``piggy_bank_part2.wasm.v1`` in the folder ``./target/concordium/wasm32-unknown-unknown/release/``.

>    Please use the wasm module file suffixed with ``v1`` when deploying to the chain. The ``piggy_bank_part2.wasm`` file (without the ``v1`` suffix) is not the ``v0`` module, but rather the raw wasm module produced by cargo.

Reviewer 2: I would recommend that the command we supply here specifies the output file
```console
cargo concordium build --out my-contract.wasm.v1
```
instead of telling them to find it among the build artifacts.

>    The ``v1`` suffixed module has been stripped for debugging information, which makes it significantly smaller. Modules already include the smart contract version (and the size of the module).

Reviewer: ```suggestion
   The ``v1`` suffixed module has been stripped for debugging information, which makes it significantly smaller.
```
The last part is not so relevant for users. I just mentioned it, because cargo-concordium v1 didn't add the smart contract version. Instead, that was added by concordium-client when deploying the module. But we changed that behaviour now, so the version is added when building with cargo-concordium to avoid problems now that we have multiple smart contract versions.

> The ``piggy_bank_part2.wasm.v1`` file can be deployed to the Concordium chain. Let’s transfer the ``piggy_bank_part2.wasm.v1`` file to the place where you are running your concordium-client tool. Move to the folder as well.

Reviewer: ```suggestion
The ``piggy_bank_part2.wasm.v1`` file can be deployed to the Concordium chain. Let’s transfer the ``piggy_bank_part2.wasm.v1`` file to the place where you are running your ``concordium-client`` tool. Move to the folder as well.
```

Deploying a module
==================
You are now set up to deploy the piggy bank module to the testnet chain. Let's give the module the name ``piggy_bank_part2_module`` so it can be referenced in this tutorial.

.. Note::
   Please remember to adjust the following commands when you are running your ``concordium-client`` tool locally and want to connect to your node on the server. You can look up the different options on how to connect the ``concordium-client`` tool to your testnet node :ref:`here<interacting-with-your-testnet-node>`.

.. code-block:: console

>    $./concordium-client module deploy piggy_bank_part2.wasm.v1 --sender <account-name> --name piggy_bank_part2_module --grpc-port 10001

Reviewer: We can mention that the names are only stored locally in the concordium client configuration. They are not visible to others on the chain. So they only exist as a convenience to the user 😊 
(That is the case for all the names/aliases, i.e., contract, module, and account names)

>    $./concordium-client module deploy piggy_bank_part2.wasm.v1 --sender <account-name> --name piggy_bank_part2_module --grpc-port 10001

Reviewer 2: I think we should promote the convention that "network configuration" comes first

```console
./concordium-client --grpc-port 10001 module deploy piggy_bank_part2.wasm.v1 --sender <account-name> --name piggy_bank_part2_module 
```

Author: In my opinion, I would like to keep the main parts of the commands that are changing as far as possible to the left, so people can see that part easily. Everything that is towards the end of a command can only be seen when operating the scroll bar so I would like to not have the important parts there. I also think that most people will copy-paste the given commands found in the previous chapters and just at the ``--grpc-port 10001`` at the end of the command when executing it. Let me know what you think.

Reviewer 2: I see.

I still prefer the network stuff first, and that is also how the command line parser is designed. But I don't feel strongly about it, so if you think this is better I'm OK with it.

Author: Thanks. I would prefer the current order for readability.

.. Note::
   This tutorial uses ``Concordium`` for the <account-name>.

.. Note::
   You will be asked to input a password. Use the ``export password`` that you used for creating the key backup file.

If everything works correctly the output has a green line with your module reference. You can find additional information about deploying a module in this :ref:`guide<deploy-module>`.

.. code-block:: console

   Module successfully deployed with reference: <module-reference>.

>    Module reference <module-reference> was successfully named 'piggy_bank_part2_module'.

Reviewer: We should mention that this will fail if a module with the exact same hash (which is also the reference) has already been deployed.
If you compile the same contract on different systems or versions of `rustc`, then their hashes will often differ.

If they _do_ hit that issue, then the module reference will be printed and you can use it for continuing this tutorial.
Something like:
`Module with reference <mod-ref> already exists on chain..`
`concordium-client module name <mod-ref> --name piggy_bank_part2_module`
And then they can continue as if they deployed it themselves :blush:

Initializing a module
=====================

> You can initialize your smart contract now to create a smart contract instance. Let's give the instance the name ``piggy_bank_part2_instance`` so it can be referenced in this tutorial.

Reviewer 3: ```suggestion
You can initialize your smart contract now to create a smart contract instance. Give the instance the name ``piggy_bank_part2_instance`` so it can be referenced in this tutorial.
```

.. code-block:: console

   $./concordium-client contract init piggy_bank_part2_module --sender <account-name> --contract PiggyBank --name piggy_bank_part2_instance --energy 1000 --grpc-port 10001

.. Note::

>    1000 is enough energy to initialize the piggy bank smart contract but you might need more energy for larger smart contracts.

Reviewer: I would also add something like: `The --energy parameter sets the maximum allowed energy to be used. You only pay for the actual energy used.`

> If everything works correctly the output has a green line with your smart contract instance index.

Reviewer 2: Could we show a screenshot?

.. code-block:: console

   Contract successfully initialized with address: {"index":<smart-contract-instance-index>,"subindex":0}
   Contract address {"index":<smart-contract-instance-index>,"subindex":0} was successfully named 'piggy_bank_part2_instance'.

You can find additional information about initializing a smart contract instance in this :ref:`guide<initialize-contract>`.

Updating the piggy bank smart contract (sending CCD)
====================================================

> Let's deposit 1 CCD into the piggy bank to test if you can update the smart contract instance.

Reviewer 3: ```suggestion
Deposit 1 CCD into the piggy bank to test if you can update the smart contract instance.
```

.. code-block:: console

   $./concordium-client contract update piggy_bank_part2_instance --entrypoint insert --energy 1000 --sender <account-name> --amount 1 --grpc-port 10001

If everything works correctly the output has a green line as follows:

.. code-block:: console

   Successfully updated contract instance {"index":<smart-contract-instance-index>,"subindex":0} ('piggy_bank_part2_instance') using the function 'insert'.

You can find additional information about updating a smart contract instance in this :ref:`guide<interact-instance>`.

Viewing the piggy bank state
============================

You can display the raw return value of your piggy bank smart contract with the following command.

.. code-block:: console

   $./concordium-client contract invoke piggy_bank_part2_instance --entrypoint view --energy 1000 --grpc-port 10001

If everything works correctly the output has a green line as follows:

.. code-block:: console

   Invocation resulted in success:
   - Energy used: 510 NRG
   - Return value (raw):
   [0,64,66,15,0,0,0,0,0]
   .

You can find additional information about invoking in this :ref:`guide<invoke-instance>`.

> Updating the piggy bank smart contract (sending no CCD)

Reviewer: With the comment from my previous review, I didn't mean that this section should have such a focus on the `no CCD` part. I thought of it as a side note that would illustrate the difference between calling payable and non-payable entrypoints. I thought the primary focus of this section would be to show the smash entrypoint, which is the only entrypoint not already covered.

But anyhow, seeing it now, I don't think this section adds so much value to the tutorial. You can remove if it you want, or, alternatively, change the title to be smashing focused rather than `sending no CCD`.

Reviewer 2: I think the second option is better. With an explicit point about payable/non-payable and what the response would be if they mixed it up.

=======================================================

> Let's smash the piggy bank contract to test that you can update the smart contract instance without sending CCD.

Reviewer 3: ```suggestion
Smash the piggy bank contract to test that you can update the smart contract instance without sending CCD.
```

.. code-block:: console

   $./concordium-client contract update piggy_bank_part2_instance --entrypoint smash --sender <account-name> --energy 2000 --grpc-port 10001

.. Note::

>    Ensure that you don't send any CCD to the smart contract instance and that the <account-name> is the same as the <account-name> that initialized the piggy bank smart contract instance.

Reviewer: If you keep this section, then you should mention that this is because the smash entrypoint isn't payable and link to the section where we describe that attribute: https://developer.concordium.software/en/mainnet/smart-contracts/tutorials/piggy-bank/writing.html#inserting-ccd

> If everything works correctly the output has a green line as follows:

Reviewer 2: And the same here, about a screenshot.

.. code-block:: console

   Successfully updated contract instance {"index":<smart-contract-instance-index>,"subindex":0} ('piggy_bank_part2_instance') using the function 'smash'.

You can find additional information about updating a smart contract instance in this :ref:`guide<interact-instance>`.

> Congratulation. You went through the whole piggy bank tutorial.

Reviewer 3: ```suggestion
Congratulations. You have completed the whole piggy bank tutorial.
```

This part focuses on how you can deploy your developed piggy bank smart contract to the Concordium testnet and interact with it.

.. warning::

   The reader is assumed to have basic knowledge of what a blockchain and smart
   contract is, and some experience with Rust_.

Preparation
===========

Before you start, make sure you completed the previous parts of the tutorial and have a running testnet node that has caught up to the head of the blockchain and an account with some CCD setup.

Building a Wasm module
======================

You have already built a Wasm module from your piggy bank smart contract in a :ref:`previous part<cargo-concordium-build>` of this tutorial.

Repeat this step since you might have done some changes to the smart contract code in :ref:`part 2<piggy-bank-testing>` of this tutorial. Go to the folder where you saved your piggy bank smart contract and use ``cargo concordium build`` to create a Wasm module.

.. code-block:: console

   $cargo concordium build --out piggy_bank_part2.wasm.v1

.. note::

   This command outputs the module ``piggy_bank_part2.wasm.v1`` to the project's root folder.

.. note::

>    When you use the above command without the ``--out`` flag, you can find the module in the folder ``./target/concordium/wasm32-unknown-unknown/release/``.

Reviewer: ```suggestion
   When you use the above command *without* the ``--out`` flag, you can find the module in the folder ``./target/concordium/wasm32-unknown-unknown/release/``.
```

>    Use the wasm module file suffixed with ``v1`` when deploying to the chain. The ``piggy_bank_part2.wasm`` file (without the ``v1`` suffix) is not the ``v0`` module, but rather the raw wasm module produced by cargo.

Reviewer 2: I think if you supply `--out` then the `.v1` file will not be in `./target/concordium/wasm32-unknown-unknown/release`.

Author: I updated the sentence to:
`When you use the above command without the ``--out`` flag, you can find the module ... `

Deploying a module
==================
You are now set up to deploy the piggy bank module to the testnet chain. Give the module the name ``piggy_bank_part2_module`` so it can be referenced in this tutorial.

.. note::

   Remember to adjust the following commands when you are running your ``concordium-client`` tool locally and want to connect to your node on the server. You can look up the different options on how to connect the ``concordium-client`` tool to your testnet node :ref:`here<interacting-with-your-testnet-node>`.

.. note::

   Names are only stored locally in the ``concordium client`` configuration.
   They are not visible to others on the chain. This is the case for all contracts, modules, and account names/aliases.

.. code-block:: console

   $./concordium-client module deploy piggy_bank_part2.wasm.v1 --sender <account-name> --name piggy_bank_part2_module --grpc-port 10001

.. note::

   This tutorial uses ``Concordium`` for the <account-name>.

.. note::

>    You will be asked to input a password. Use the ``export password`` that you used to create the key backup file.

Reviewer: .. in part 3 of the piggy bank tutorial (add link).

.. code-block:: console

   Invocation resulted in success:
   - Energy used: 510 NRG
   - Return value (raw):
   [0,64,66,15,0,0,0,0,0]
   .

.. image:: ./images/pb_tutorial_24.png
   :width: 100 %

You can find additional information about invoking in this :ref:`guide<invoke-instance>`.

Smashing the piggy bank smart contract
======================================

Smash the piggy bank contract to test that you can invoke this entry point. After smashing the piggy bank you cannot enter any additional CCD into the piggy bank anymore.

.. code-block:: console

   $./concordium-client contract update piggy_bank_part2_instance --entrypoint smash --sender <account-name> --energy 2000 --grpc-port 10001

If everything works correctly the output has a green line as follows:

.. code-block:: console

   Successfully updated contract instance {"index":<smart-contract-instance-index>,"subindex":0} ('piggy_bank_part2_instance') using the function 'smash'.

.. image:: ./images/pb_tutorial_25.png
   :width: 100 %

You can find additional information about updating a smart contract instance in this :ref:`guide<interact-instance>`.

> .. note::

Reviewer: Nice note. I think this clarifies a few concepts ;)

   Ensure that you don't send any CCD to the smart contract instance and that the ``--sender`` flag uses the <account-name>
   that initialized the piggy bank smart contract instance (the owner of this smart contract instance).
   You can find explanations about the behavior of the ``Smash`` entry point in this :ref:`guide<smashing-the-piggy-bank-writing>`.
   The ``Smash`` entry point is ``non-payable`` and will print the below error message if you send CCD with the ``--amount`` flag.
   The opposite is not true and you can omit the ``--amount`` flag when interacting
   with a ``payable`` function without raising an error because in that case 0 CCD is sent to the ``payable`` entry point.

.. code-block:: console

   Error: Updating contract instance failed:

>    'smash' in 'PiggyBank' at {"index":<smart-contract-instance-index>,"subindex":0} failed with code -2147483636.

Reviewer: Please write that they can find the meaning of common error codes by looking here: https://docs.rs/concordium-std/3.0.0/concordium_std/#signalling-errors

(The errors from concordium-std start from `i32::MIN` and go upwards. The contract's own error types start from `-1` and go downwards.)

You can find additional information about updating a smart contract instance in this :ref:`guide<interact-instance>`.

``Concordium-std`` crate errors
-------------------------------

> .. note::

Reviewer: I don't think the first paragraph in each of these error sections should be notes. I find it odd to have a section that is mostly side-notes.
So just remove the `.. note::` line for the first paragraph in the two error sections 😊

.. image:: ./images/pb_tutorial_21.png
   :width: 100 %

.. note::
   You can find the meaning of common error codes in this `list <https://docs.rs/concordium-std/3.0.0/concordium_std/#signalling-errors>`_.
   This error was thrown by the ``concordium-std`` crate. The error codes start from ``i32::MIN`` and go upwards.
   ``-2147483636`` corresponds to ``NotPayableError``. In contrast, error types of the smart contract start from ``-1`` and go downwards.

Smart contract errors
---------------------

> .. note::

Reviewer: ```suggestion
```

   Ensure that the ``--sender`` flag uses the <account-name>

>    that initialized the piggy bank smart contract instance (the owner of this smart contract instance).

Reviewer: Please mention that it will result in the error shown below. Similar to what you do in the first error section.
